private/pkg/transport/http/httpserver/runner.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpserver
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/bufbuild/buf/private/pkg/rpc/rpchttp"
+	"github.com/go-chi/chi/v5"
+	"github.com/go-chi/chi/v5/middleware"
+	"go.uber.org/zap"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
+	"golang.org/x/sync/errgroup"
+)
+
+type runner struct {
+	logger            *zap.Logger
+	shutdownTimeout   time.Duration
+	readHeaderTimeout time.Duration
+	tlsConfig         *tls.Config
+	middlewares       []func(http.Handler) http.Handler
+	health            bool
+	maxBodySize       int64
+	walkFunc          chi.WalkFunc
+}
+
+func newRunner(logger *zap.Logger, options ...RunnerOption) *runner {
+	runner := &runner{
+		logger:            logger.Named("httpserver"),
+		shutdownTimeout:   DefaultShutdownTimeout,
+		readHeaderTimeout: DefaultReadHeaderTimeout,
+	}
+	for _, option := range options {
+		option(runner)
+	}
+	return runner
+}
+
+func (s *runner) Run(
+	ctx context.Context,
+	listener net.Listener,
+	mappers ...Mapper,
+) (retErr error) {
+	start := time.Now()
+	defer func() {
+		if retErr != nil {
+			s.logger.Error("finished", zap.Duration("duration", time.Since(start)), zap.Error(retErr))
+		} else {
+			s.logger.Info("finished", zap.Duration("duration", time.Since(start)))
+		}
+	}()
+
+	mux := chi.NewMux()
+	mux.Use(middleware.Recoverer)
+	mux.Use(middleware.StripSlashes)
+	mux.Use(newZapMiddleware(s.logger))
+	if s.maxBodySize > 0 {
+		mux.Use(func(next http.Handler) http.Handler {
+			return http.MaxBytesHandler(next, s.maxBodySize)
+		})
+	}
+	mux.Use(s.middlewares...)
```

Comment:
Should user-defined middleware _always_ come last, maybe move this one line below?